pyrometheus/codegen/python.py

```diff
@@ -286,6 +293,69 @@ def get_mixture_internal_energy_mass(self, temperature, mass_fractions):
+        return self.usr_np.sqrt(temperature)*self._pyro_make_array([
+                % for sp in range(sol.n_species):
+                ${cgm(ce.transport_polynomial_expr(sol.get_viscosity_polynomial(sp),
+                                                    2, Variable("temperature")))},
+                % endfor
+                ])
+
+    def get_mixture_viscosity_mixavg(self, temperature, mass_fractions):
+        mole_fractions = self.iwts * mass_fractions\
+ * self.get_mix_molecular_weight(mass_fractions)
+        viscosities = self.get_species_viscosities(temperature)
+        mix_rule_f = self._pyro_make_array([
+            %for sp in range(sol.n_species):
+            ${cgm(ce.viscosity_mixture_rule_wilke_expr(sol, sp,
+                 Variable("mole_fractions"), Variable("viscosities")))},
+            %endfor
+            ])
+        return sum(mole_fractions*viscosities/mix_rule_f)
+
+    def get_species_thermal_conductivities(self, temperature):
+        return self.usr_np.sqrt(temperature)*self._pyro_make_array([
+                % for sp in range(sol.n_species):
+                ${cgm(ce.transport_polynomial_expr(
+                    sol.get_thermal_conductivity_polynomial(sp), 1,
+                    Variable("temperature")))},
+                % endfor
+                ])
+
+    def get_mixture_thermal_conductivity_mixavg(self, temperature, mass_fractions):
+        mole_fractions = self.iwts * mass_fractions\
+ * self.get_mix_molecular_weight(mass_fractions)
+        conductivities = self.get_species_thermal_conductivities(temperature)
+        return 0.5*(sum(mole_fractions*conductivities)
+            + 1/sum(mole_fractions/conductivities))
+
+    def get_species_binary_mass_diffusivities(self, temperature):
+        return self._pyro_make_array([
+                % for i, j, in product(range(sol.n_species), range(sol.n_species)):
+                ${cgm(ce.transport_polynomial_expr(
+                      sol.get_binary_diff_coeffs_polynomial(i, j), 1,
+                      Variable("temperature")))},
+                % endfor
+                ]).reshape((self.num_species, self.num_species))
+
+    def get_species_mass_diffusivities_mixavg(self, temperature, pressure,
+                                              mass_fractions, threshold=1.e-7):
+        mmw = self.get_mix_molecular_weight(mass_fractions)
+        mole_fractions = self.iwts * mass_fractions * mmw
+        diff_ij = self.get_species_binary_mass_diffusivities(temperature)
+        mix_rule_f = self._pyro_make_array([
+              % for sp in range(sol.n_species):
+              self.usr_np.where(
+                  self.usr_np.greater(1.0-mole_fractions[${sp}], threshold),
+                  ${cgm(ce.species_mass_diff_mixture_rule_expr(sol, sp,
+                      Variable("mmw"), Variable("mole_fractions"),
+                      Variable("diff_ij")))},
+                  diff_ij[${cgm(sp)}, ${cgm(sp)}],
```

Comment:
This seems... inelegant. Is there a better way to do this?

Reply:
The "where" is necessary to switch from a mixture-rule to the self-diffusivity and avoid a singularity of 0/0. Tests with Mirge-com result in NaN in the very first iteration when not considering the self-diffusivity. In my opinion, it is better to check here which coefficient to use, instead of returning both D_i^m (mixture) and D_ii (self) to MIRGE-COM and do the checking over there.

If the comparison is done against a hard 1.0 (for instance "self.usr_np.less(mole_fractions[${sp}], 1.0)"), without considering round-off errors, it is possible that the code fails. For instance, the output of "print(dii_pm[i], dii_ct[i], y)", which is line 217 in the test_codegen.py results in:

5.092329409340231e-06 5.092329409340232e-06 [1. 0. 0. 0. 0. 0. 0.]
9.923190033460186e-06 9.923190033460188e-06 [0. 1. 0. 0. 0. 0. 0.]
4.869780035934058e-06 4.869780035934059e-06 [0. 0. 1. 0. 0. 0. 0.]
9.842024372037339e-06 9.842024372037342e-06 [0. 0. 0. 1. 0. 0. 0.]
8.234238345402362e-06 8.234238345402364e-06 [0. 0. 0. 0. 1. 0. 0.]
  7.37190664636879e-05 7.371906646368791e-05 [0. 0. 0. 0. 0. 1. 0.]
                                  nan 1.002002531371639e-05 [0. 0. 0. 0. 0. 0. 1.]
                                  
However, considering even a round-off error of 1e-15 fixed the problem and the code doesn't fail/crash. I added the "threshold" option to give the user the decision to consider when to switch from a mixture-rule to self diffusivity. This could be passed down from the final application (Mirge-COM, in my case) towards this function.

Please, let me know your thoughts on this.